clients/src/test/java/org/apache/kafka/common/record/LazyDownConversionRecordsTest.java

```diff
@@ -123,50 +55,138 @@ public void doTestConversion() throws IOException {
             File outputFile = tempFile();
             FileChannel channel = new RandomAccessFile(outputFile, "rw").getChannel();
 
-            // Size of lazy records is at least as much as the size of underlying records
-            assertTrue(lazyRecords.sizeInBytes() >= inputRecords.sizeInBytes());
+            int written = 0;
+            while (written < toWrite)
+                written += lazySend.writeTo(channel, written, toWrite - written);
 
+            FileRecords convertedRecords = FileRecords.open(outputFile, true, (int) channel.size(), false);
+            ByteBuffer convertedRecordsBuffer = ByteBuffer.allocate(convertedRecords.sizeInBytes());
+            convertedRecords.readInto(convertedRecordsBuffer, 0);
+
+            // cleanup
+            convertedRecords.close();
+            channel.close();
+
+            return MemoryRecords.readableRecords(convertedRecordsBuffer);
+        }
+    }
+
+    @Test
+    public void testConversionOfCommitMarker() throws IOException {
+        MemoryRecords recordsToConvert = MemoryRecords.withEndTransactionMarker(0, Time.SYSTEM.milliseconds(), RecordBatch.NO_PARTITION_LEADER_EPOCH,
+                1, (short) 1, new EndTransactionMarker(ControlRecordType.COMMIT, 0));
+        MemoryRecords convertedRecords = convertRecords(recordsToConvert, (byte) 1, recordsToConvert.sizeInBytes());
+        ByteBuffer buffer = convertedRecords.buffer();
+        buffer.getLong();
+        int sizeOfConvertedRecords = buffer.getInt();
+        assertTrue(sizeOfConvertedRecords > buffer.limit());
+        assertFalse(convertedRecords.batchIterator().hasNext());
+    }
+
+    @RunWith(value = Parameterized.class)
+    public static class ParameterizedConversionTest {
+        private final CompressionType compressionType;
+        private final byte toMagic;
+        private final DownConversionTest test;
+
+        public ParameterizedConversionTest(CompressionType compressionType, byte toMagic, DownConversionTest test) {
+            this.compressionType = compressionType;
+            this.toMagic = toMagic;
+            this.test = test;
+        }
+
+        enum DownConversionTest {
```

Comment:
I think it would be more straightforward to have two separate test cases which call `doTestConversion` with a boolean parameter to indicate whether overflow should be tested.